File: typescript/packages/node/src/core/email/read.ts
// ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
// ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========

import type { IndexCacheStore, PathSpec } from '@struktoai/mirage-core'
import type { EmailAccessor } from '../../accessor/email.ts'
import { fetchAttachment, fetchMessage } from './_client.ts'
import { rstripSlash, stripSlash } from '@struktoai/mirage-core'

const ENC = new TextEncoder()

function enoent(p: string): Error {
  const e = new Error(`ENOENT: ${p}`) as Error & { code: string }
  e.code = 'ENOENT'
  return e
}

function eisdir(p: string): Error {
  const e = new Error(`EISDIR: ${p}`) as Error & { code: string }
  e.code = 'EISDIR'
  return e
}

function dirname(p: string): string {
  const norm = rstripSlash(p)
  const idx = norm.lastIndexOf('/')
  if (idx <= 0) return '/'
  return norm.slice(0, idx)
}

export async function read(
  accessor: EmailAccessor,
  path: PathSpec,
  index?: IndexCacheStore,
): Promise<Uint8Array> {
  const prefix = path.prefix
  let p = path.original
  if (prefix !== '' && p.startsWith(prefix)) p = p.slice(prefix.length) || '/'
  const key = stripSlash(p)
  if (index === undefined) throw enoent(path.original)
  const virtualKey = prefix !== '' ? `${prefix}/${key}` : `/${key}`
  const result = await index.get(virtualKey)
  if (result.entry === undefined || result.entry === null) throw enoent(path.original)
  const rt = result.entry.resourceType
  if (rt === 'email/folder' || rt === 'email/date' || rt === 'email/attachment_dir') {
    throw eisdir(path.original)
  }
  if (rt === 'email/attachment') {
    const parentKey = dirname(virtualKey)
    const parentResult = await index.get(parentKey)
    if (parentResult.entry === undefined || parentResult.entry === null) {
      throw enoent(path.original)
    }
    const uid = parentResult.entry.id
    const parts = stripSlash(virtualKey).split('/')
    const folder = prefix !== '' ? (parts[1] ?? '') : (parts[0] ?? '')
    const filename = result.entry.vfsName !== '' ? result.entry.vfsName : result.entry.name
    const data = await fetchAttachment(accessor, folder, uid, filename)
    if (data === null) throw enoent(path.original)
    return data
  }
  const parts = stripSlash(virtualKey).split('/')
  const folder = prefix !== '' ? (parts[1] ?? '') : (parts[0] ?? '')
  const uid = result.entry.id
  const msg = await fetchMessage(accessor, folder, uid)
  return ENC.encode(JSON.stringify(msg))
}

File: python/mirage/core/email/read.py
# ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
# ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========

import json
import posixpath

from mirage.accessor.email import EmailAccessor
from mirage.cache.index import IndexCacheStore
from mirage.core.email._client import fetch_attachment, fetch_message
from mirage.types import PathSpec
from mirage.utils.errors import enoent


async def read(
    accessor: EmailAccessor,
    path: PathSpec,
    index: IndexCacheStore = None,
) -> bytes:
    if isinstance(path, str):
        path = PathSpec(original=path, directory=path)
    virtual = path.original
    if isinstance(path, PathSpec):
        prefix = path.prefix
        path = path.original

    if prefix and path.startswith(prefix):
        rest = path[len(prefix):]
        if prefix.endswith("/") or rest == "" or rest.startswith("/"):
            path = rest or "/"
    key = path.strip("/")
    if index is None:
        raise enoent(virtual)
    virtual_key = prefix + "/" + key if prefix else "/" + key
    result = await index.get(virtual_key)
    if result.entry is None:
        raise enoent(virtual)
    if result.entry.resource_type in ("email/folder", "email/date",
                                      "email/attachment_dir"):
        raise IsADirectoryError(virtual)
    if result.entry.resource_type == "email/attachment":
        parent_key = posixpath.dirname(virtual_key)
        parent_result = await index.get(parent_key)
        if parent_result.entry is None:
            raise enoent(virtual)
        uid = parent_result.entry.id
        parts = virtual_key.strip("/").split("/")
        folder = parts[1] if prefix else parts[0]
        filename = result.entry.vfs_name
        data = await fetch_attachment(accessor, folder, uid, filename)
        if data is None:
            raise enoent(virtual)
        return data
    parts = virtual_key.strip("/").split("/")
    folder = parts[1] if prefix else parts[0]
    uid = result.entry.id
    msg = await fetch_message(accessor, folder, uid)
    return json.dumps(msg, ensure_ascii=False).encode()
